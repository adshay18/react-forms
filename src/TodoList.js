import { useState } from 'react';
import Todo from './Todo';
import { v4 as uuid } from 'uuid';
import NewTodoForm from './NewTodoForm';

const TodoList = () => {
	const [ todos, setTodos ] = useState([]);

	const addTodo = (todo) => {
		let newTodo = { ...todo, id: uuid() };
		setTodos((todos) => [ ...todos, newTodo ]);
	};

	const removeTodo = (idx) => {
		let todosCopy = [ ...todos ];
		let todo = todosCopy.splice(todosCopy[idx], 1);
		setTodos((todos) => [ ...todosCopy ]);
	};

	return (
		<div className="TodoList">
			<NewTodoForm addTodo={addTodo} />
			<h2>Todos:</h2>
			{todos.map((todo, i) => <Todo key={todo.id} idx={i} removeTodo={removeTodo} text={todo.text} />)}
		</div>
	);
};

export default TodoList;
