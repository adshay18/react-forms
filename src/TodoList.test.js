import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';

// Smoke test
it('renders without crashing', function() {
	render(<TodoList />);
});

// Snapshot
it('matches snapshot', function() {
	const { asFragment } = render(<TodoList />);
	expect(asFragment()).toMatchSnapshot();
});

it('can add a new todo and delete it', function() {
	const { getByLabelText, queryByText } = render(<TodoList />);

	// no items yet
	expect(document.querySelector('div.Todo')).not.toBeInTheDocument();

	const textInput = getByLabelText('Todo:');
	const submitBtn = queryByText('Add Task');

	// fill out the form
	fireEvent.change(textInput, { target: { value: 'Test todo' } });
	fireEvent.click(submitBtn);

	// item exists!
	expect(document.querySelector('div.Todo')).toBeInTheDocument();

	const deleteBtn = queryByText('Done');

	// click delete button
	fireEvent.click(deleteBtn);

	// box is deleted
	expect(document.querySelector('div.Todo')).not.toBeInTheDocument();
});
