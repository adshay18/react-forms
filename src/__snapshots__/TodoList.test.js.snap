// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <form>
      <label
        for="text"
      >
        Todo: 
      </label>
      <input
        id="text"
        name="text"
        type="text"
        value=""
      />
      <button>
        Add Task
      </button>
    </form>
    <h2>
      Todos:
    </h2>
  </div>
</DocumentFragment>
`;
